business/authentication/openid_auth_controller.go

```diff
@@ -137,41 +134,15 @@ func NewOpenIdAuthController(persistor SessionPersistor, businessInstantiator fu
 	}
 }
 
-// Authenticate is the entry point to handle OpenId authentication using the implicit flow. The HTTP request
-// should contain "id_token" and "state" as URL parameters. If RBAC is enabled, the id_token should be
-// valid to be used in the Kubernetes API (thus, privileges are verified to allow login); else, only token
-// validity is checked and users will share the same privileges.
-// An AuthenticationFailureError is returned if the authentication failed. Any
-// other kind of error means that something unexpected happened.
+// Authenticate was the entry point to handle OpenId authentication using the implicit flow. Support
+// for the implicit flow has been removed. This is left here, because the "Authenticate" function is required
+// by the AuthController interface which must be implemented by all auth controllers. So, this simply
+// returns an error.
```

Comment:
Auth is now just handled by providing a redirect url?

Reply:
I would say "no"?
Although, I think I don't fully understand the question...

Reply:
I was wondering why the `Authenticate` function is no longer used with just the code flow?

Reply:
Well, in OpenID, we tell in the docs to configure the IdP to callback to the Kiali "web root". IIRC, this is because we started with the implicit flow.

In the implicit flow, when the IdP calls back to the "web root" the front-end can boot and dispatch the login flow (i.e. it collects callback params and fires a request to `api/authenticate` using those params). The `Authenticate` func runs when the `api/authenticate` endpoint is hit. This is done this way because the IdP will callback with _hash parameters_ which can only be read by the front-end and aren't accessible in the back-end. So, the help of the front-end was needed to pass parameters to the back-end through the right method.

Then, we implemented the code flow. I no longer remember correctly, but perhaps for consistency (and simpler docs) we chose to keep the need to configure the IdP to callback the "web root". Otherwise, people would need to know which flow will become into effect to set the right callback endpoint in the IdP. However, in the code flow there is no need to rely in the UI and the back-end can handle everything alone. So, auth is handled as soon as the IdP calls back but since it is not a callback to the `api/authenticate` the go-router won't call the `Authenticate` function. For this, the router is patched here: https://github.com/kiali/kiali/blob/f3ef5f41b9a4c45aba27a143d0173b781d4980fe/routing/router.go#L102-L107.

Honestly, nowadays  I would prefer to have a dedicated callback endpoint for OpenID (and also for OpenShift). However, that would be a breaking change as everybody would need to re-configure their IdPs. So, I'm simply keeping things backwards compatible.

Not sure if this makes sense, or if I'm confusing you more :smile: